src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php

```diff
@@ -48,45 +50,89 @@ public function updatePositionsAction(Request $request): Response
                 } catch (\InvalidArgumentException $exception) {
                     throw new HttpException(Response::HTTP_BAD_REQUEST, $exception->getMessage());
                 }
-
-                $this->manager->flush();
             }
         }
 
         return new JsonResponse();
     }
 
-    /**
-     * @psalm-suppress DeprecatedMethod
-     */
     public function updateProductTaxonsPositionsAction(Request $request): Response
     {
         $configuration = $this->requestConfigurationFactory->create($this->metadata, $request);
+        $this->validateCsrfProtection($request, $configuration);
         $this->isGrantedOr403($configuration, ResourceActions::UPDATE);
-        $productTaxons = $this->getParameterFromRequest($request, 'productTaxons');
+        $productTaxonsPositions = $request->request->all('productTaxons');
 
-        $this->validateCsrfProtection($request, $configuration);
+        if (!$this->shouldProductsPositionsBeUpdated($request, $productTaxonsPositions)) {
+            return $this->redirectHandler->redirectToReferer($configuration);
+        }
 
-        if ($this->shouldProductsPositionsBeUpdated($request, $productTaxons)) {
+        try {
+            $productTaxonsPositionsMap = $this->mapFromArrayToProductTaxonPositionMap($productTaxonsPositions);
+        } catch (\InvalidArgumentException $exception) {
             /** @var Session $session */
             $session = $request->getSession();
+            $session->getFlashBag()->add('error', $exception->getMessage());
 
-            foreach ($productTaxons as $id => $position) {
-                try {
-                    $this->updatePositions($position, $id);
-                } catch (\InvalidArgumentException $exception) {
-                    $session->getFlashBag()->add('error', $exception->getMessage());
+            return $this->redirectHandler->redirectToReferer($configuration);
+        }
 
-                    return $this->redirectHandler->redirectToReferer($configuration);
-                }
-            }
+        /** @var PositionerInterface $positioner */
+        $positioner = $this->get(PositionerInterface::class);
 
+        foreach ($productTaxonsPositionsMap as $productTaxonPositionMap) {
+            $positioner->updatePosition(
+                $productTaxonPositionMap['productTaxon'],
+                $productTaxonPositionMap['newPosition'],
+                $this->getMaxPosition(),
```

Comment:
Extract getting max position before this foreach not to query the database many times